### PR TITLE
UI to add support users to the database

### PR DESCRIPTION
**Pull request by duncanjbrown**
### Context

To support #770, create a UI to add support users to the db

- we should not merge #770 until this is merged and we have created at least one support user, as without a user in the database it will be impossible to log into support once basic auth is turned off

### Changes proposed in this pull request

CRUD for `SupportUsers` in the support interface.

- We don't support deletion as we haven't discussed how this will work
- at the moment all support users will be able to create other users, which is a bit 😬, but we can revisit

### Link to Trello card

https://trello.com/c/Yk89F38A/1371-ui-to-add-support-users-to-the-database

out

### CODE REVIEW

**Comment by tvararu** [comment]
> we should not merge #770 until this is merged and we have created at least one support user, as without a user in the database it will be impossible to log into support once basic auth is turned off

This sounds annoying when we create a new environment - is it worth exploring if we can make the current basic auth a progressive fallback or explicitly togglable via feature flag? My thinking is:

- When the db is first created, use basic auth to access support console
- Create your support users
- Pull the plug 🔌 by toggling on the feature flag: disables the basic auth mechanism, DSI only moving forward

**Comment by duncanjbrown** [comment]
> > we should not merge #770 until this is merged and we have created at least one support user, as without a user in the database it will be impossible to log into support once basic auth is turned off
> 
> This sounds annoying when we create a new environment - is it worth exploring if we can make the current basic auth a progressive fallback or explicitly togglable via feature flag? My thinking is:
> 
> * When the db is first created, use basic auth to access support console
> * Create your support users
> * Pull the plug 🔌 by toggling on the feature flag: disables the basic auth mechanism, DSI only moving forward

Good point. One alternative would be to connect directly to the db and create a user, but that's not terribly slick. Whatever we do would need to go on top of #770 and #815 — shall we revisit once those are merged?

**Comment by tvararu** [comment]
Discussed offline: because leaving in basic auth is icky, and we shouldn't have a need to create new environments too often, we think it's best if we go with the DB insert method. We should document this.

File: spec/system/support_interface/managing_support_users_spec.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    then_i_should_see_the_list_of_support_users
+    and_i_should_see_the_user_i_created
+  end
+
+  def given_i_am_a_support_user
+    page.driver.browser.authorize('test', 'test')
+  end
+
+  def and_a_support_user_exists_in_the_database
+    create(:support_user, email_address: 'person@education.gov.uk')
```
Comment (tijmenb) [inline spec/system/support_interface/managing_support_users_spec.rb]: Nooooo
```suggestion
    create(:support_user, email_address: 'person@example.com')
```

File: spec/system/support_interface/managing_support_users_spec.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def when_i_visit_the_support_console
+    visit '/support'
```
Comment (tijmenb) [inline spec/system/support_interface/managing_support_users_spec.rb]:
```suggestion
    visit support_interface_path
```

(iirc)

File: spec/system/support_interface/managing_support_users_spec.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def and_i_click_the_manange_support_users_link
+    click_link 'Support users'
+  end
+
+  def and_i_enter_the_users_email_and_dsi_uid
+    fill_in 'support_user[email_address]', with: 'harrison@bergeron.com'
```
Comment (tijmenb) [inline spec/system/support_interface/managing_support_users_spec.rb]:
```suggestion
    fill_in 'support_user[email_address]', with: 'harrison@example.com'
```

File: app/views/support_interface/support_users/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<% content_for :title, 'Add support user' %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <%= form_with model: @support_user, url: support_interface_support_users_path, builder: GOVUKDesignSystemFormBuilder::FormBuilder do |f| %>
+      <%= f.govuk_error_summary %>
```
Comment (fofr) [inline app/views/support_interface/support_users/new.html.erb]: Error summaries should always be above the `h1`.

File: spec/system/support_interface/managing_support_users_spec.rb (hunk 1 above)
Comment (duncanjbrown) [reply spec/system/support_interface/managing_support_users_spec.rb]: Ah — am I breaking with convention?

File: app/views/layouts/_support_header.html.erb [hunk 5]
```diff
@@ -6,4 +6,5 @@
   <%= nav_link 'Features', support_interface_feature_flags_path, active: 'feature_flags' %>
   <%= nav_link 'Performance', support_interface_performance_path, active: 'performance' %>
   <%= nav_link 'Tasks', support_interface_tasks_path, active: 'tasks' %>
+  <%= nav_link 'Support users', support_interface_support_users_path, active: 'users' %>
```
Comment (fofr) [inline app/views/layouts/_support_header.html.erb]: This is wrapping now.

I wonder if we should move some of these behind an "Admin" link?

Admin page would link to:
- Features
- Tasks
- Performance
- Support users

(probably not in this PR)

Comment (duncanjbrown) [reply app/views/layouts/_support_header.html.erb]: I discussed with @tvararu — we'd like to revisit this. Will loop you in @fofr when that conversation happens

Comment (tvararu) [reply app/views/layouts/_support_header.html.erb]: Actually Paul's suggestion sounds good, it was what I was thinking about too 🤔 if it's not too much, we can do it as part of this?

File: config/locales/en.yml [hunk 6]
```diff
@@ -115,6 +115,12 @@ en:
             relationship:
               blank: Tell us about the referee’s relationship to you and how long they’ve known you
               too_many_words: Relationship to referee must be %{count} words or fewer
+        support_user:
+          attributes:
+            email_address:
+              blank: Email address can’t be blank
+            dfe_sign_in_uid:
+              blank: DFE Sign-in UID can’t be blank
```
Comment (stevehook) [inline config/locales/en.yml]: 'DfE' for consistency?

File: app/views/support_interface/support_users/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+      <%= f.govuk_text_field :email_address, label: { text: 'Email address' } %>
+      <%= f.govuk_text_field :dfe_sign_in_uid, label: { text: 'DFE Sign-in UID' } %>
```
Comment (stevehook) [inline app/views/support_interface/support_users/new.html.erb]: 'DfE'?

File: config/locales/en.yml (hunk 6 above)
Comment (duncanjbrown) [reply config/locales/en.yml]: 👌

File: app/views/layouts/_support_header.html.erb (hunk 5 above)
Comment (duncanjbrown) [reply app/views/layouts/_support_header.html.erb]: I'll do another PR after this

**Comment by duncanjbrown** [comment]
Thanks everyone!

- made all the emails in specs `@example.com` (🙏  @tijmenb)
- included instructions for creating a support user in prod, and while I'm at it add some generic instructions for connecting to live databases (🙏  @tvararu and @tijmenb)
- pushed the error summary above the `h1` (🙏 @fofr)
- correctly capitalised DfE Sign-in (🙏 @stevehook)

If I may, I'll punt the menu reorg (@fofr and @tvararu) until the next PR as this is a useful change in itself and no members of the public will see the overflowing menu

**Comment by tijmenb** [comment]
Merging because I need this UI for a follow up PR!

**Pull request merged**